### PR TITLE
Fix TARGET_FILE keyword not working

### LINKED ISSUE #11135: TARGET_FILE keyword does not work
### What happened? (You can include a screenshot if it helps explain)

The TARGET_FILE keyword means we should wait for a target file for a forward model, but this does not seem to be working anymore.

### What did you expect to happen?

If the target file was specified, would expect the forward model to fail if the target file does not exist

Also:

- [ ] TARGET_FILE to be documented properly
- [ ] A good error message in the gui if the forward model failed because we did not find the target file
- [ ] Add tests

### Steps to reproduce

Edit POLY_EVAL to:
```
EXECUTABLE poly_eval.py

TARGET_FILE my_ok_file

```

Run forward model, we see target file in jobs.json:
```json
{
  "jobList": [
    {
      "name": "poly_eval",
      "executable": "...",
      "target_file": "my_ok_file",
      "error_file": null,
      "start_file": null,
      "stdout": "poly_eval.stdout.0",
      "stderr": "poly_eval.stderr.0",
      "stdin": null,
      "argList": [],
      "environment": {
        "_ERT_ITERATION_NUMBER": "0",
        "_ERT_REALIZATION_NUMBER": "1",
        "_ERT_RUNPATH": ...
      },
      "max_running_minutes": null
    }
  ],
 ```

### Environment where bug has been observed

- [ ] Python 3.11
- [ ] Python 3.12
- [ ] macOS
- [ ] RHEL8
- [ ] RHEL9
- [ ] Local queue
- [ ] LSF queue
- [ ] Slurm queue
- [ ] OpenPBS queue

### PR DESCRIPTION
**Pull request by jonathan-eq**
**Issue**
Resolves #11135

**Approach**
This commit fixes the issue where the target file is never checked in ForwardModelStep after running.

(Screenshot of new behavior in GUI if applicable)

- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Added appropriate release note label
- [x] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [ ] Make sure unit tests pass locally after every commit (`git rebase -i main
      --exec 'just rapid-tests'`)

## When applicable
- [ ] **When there are user facing changes**: Updated documentation
- [ ] **New behavior or changes to existing untested code**: Ensured that unit tests are added (See [Ground Rules](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md#ground-rules)).
- [ ] **Large PR**: Prepare changes in small commits for more convenient review
- [x] **Bug fix**: Add regression test for the bug
- [ ] **Bug fix**: Add backport label to latest release (format: 'backport release-branch-name')

### CODE REVIEW

File: src/_ert/forward_model_runner/forward_model_step.py [hunk 1]
```diff
@@ -407,8 +411,8 @@ def _assert_arg_list(self):
         return errors
 
 
-def _get_target_file_ntime(file: str | None) -> int | None:
-    mtime = None
+def _get_target_file_mtime(file: str | None) -> int:
+    mtime = 0
```
Comment (berland) [inline src/_ert/forward_model_runner/forward_model_step.py]: mtime=0 means an actual timestamp (in 1970). Isn't it better to keep `None` when we have no information on the time?

Comment (berland) [reply src/_ert/forward_model_runner/forward_model_step.py]: (nanoseconds will run out in 64 bit integers by year 2262, who knows, maybe the value will roll over to 0 😆 )

Comment (jonathan-eq) [reply src/_ert/forward_model_runner/forward_model_step.py]: Yes, we can do that too. This was just the way it was before, when it worked

Comment (jonathan-eq) [reply src/_ert/forward_model_runner/forward_model_step.py]: Imagine Ert 2262 🤔

File: src/_ert/forward_model_runner/forward_model_step.py [hunk 2]
```diff
@@ -251,8 +253,10 @@ def _create_exited_message_based_on_exit_code(
                 f"Found the error file:{self.step_data['error_file']} - step failed."
             )
 
-        if target_file_mtime:
-            target_file_error = self._check_target_file_is_written(target_file_mtime)
+        if target_file:
```
Comment (berland) [inline src/_ert/forward_model_runner/forward_model_step.py:262]: This if-statement, or the if-statement at the start of _check_target_file_is_written is testing the same, so one can be removed I think.

File: tests/ert/unit_tests/forward_model_runner/test_forward_model_step.py [hunk 3]
```diff
@@ -295,3 +295,78 @@ def test_processtree_timer(
...
+@pytest.mark.usefixtures("use_tmpdir")
+def test_target_file_specified():
+    target_file = "some_file"
+    fmstep = ForwardModelStep(
+        {
+            "name": "exit 0 and touch target file",
+            "executable": "/bin/sh",
+            "stdout": "exit_out",
+            "stderr": "exit_err",
+            "argList": ["-c", f"touch {target_file}; exit 0"],
```
Comment (berland) [inline tests/ert/unit_tests/forward_model_runner/test_forward_model_step.py]: we should probably also test this with nonzero exit code. Maybe we need to dig in the source to determine what is the correct behaviour in that scenario. @oyvindeide , you found a version of Ert that did not have this bug where we can test behaviour?

File: src/_ert/forward_model_runner/forward_model_step.py [hunk 4]
```diff
@@ -167,8 +168,8 @@ def _run(self) -> Generator[Start | Exited | Running]:
         (stdin, stdout, stderr) = self._open_file_handles()
         # stdin/stdout/stderr are closed at the end of this function
 
-        target_file = self.step_data.get("target_file")
-        target_file_mtime: int | None = _get_target_file_ntime(target_file)
+        target_file: str | None = self.step_data.get("target_file")
+        target_file_mtime: int | None = _get_target_file_mtime(target_file)
```
Comment (berland) [inline src/_ert/forward_model_runner/forward_model_step.py]: The variable naming is here a little bit confusing, because `target_file_mtime` will only have a value if the file already exists for whatever reason, typically an earlier (failed) run. (It could make sense also to just delete the target_file if it exists prior to running the step - but that is out-of-scope for this bugfix)

Comment (jonathan-eq) [reply src/_ert/forward_model_runner/forward_model_step.py]: I can rename it to `existing_target_file_mtime` if that makes it better.

File: tests/ert/unit_tests/forward_model_runner/test_forward_model_step.py (hunk 3 above)
Comment (jonathan-eq) [reply tests/ert/unit_tests/forward_model_runner/test_forward_model_step.py]: The old behavior was only for zero exitcode prior to 4dc894ca63687476e091f582df5a42045190f7bd. Should we keep it that way for now, and improve it in a different PR if necessary?

Comment (berland) [reply tests/ert/unit_tests/forward_model_runner/test_forward_model_step.py]: We keep the prior behaviour for now (and lock it down through tests)

Comment (jonathan-eq) [reply tests/ert/unit_tests/forward_model_runner/test_forward_model_step.py]: Yes, I will parametrize the tests with exit_code 0 and 1, so we also test that in all scenarios.

**Comment by jonathan-eq** [comment]
> You can cherry-pick #11144 and probably modify it to mention the behaviour for exit-code 1 + target_file.

I will document that it is only used with zero return code

File: tests/ert/unit_tests/forward_model_runner/test_forward_model_step.py [hunk 5]
```diff
@@ -295,3 +295,104 @@ def test_processtree_timer(
...
         timer.update(snapshot)
     assert timer.total_cpu_seconds() == expected_total_seconds
+
+
+@pytest.mark.usefixtures("use_tmpdir")
+@pytest.mark.parametrize(
+    "exit_code, expected_error_message",
+    [(0, None), (1, "Process exited with status code 1")],
+)
+def test_target_file_specified(exit_code, expected_error_message):
```
Comment (berland) [inline tests/ert/unit_tests/forward_model_runner/test_forward_model_step.py]: Can this test name be more explicit about what it is testing? 
Are we testing propagation of error messages, or how Ert handles the step when the target file is specified and is present, or both?

Are we testing that Ert is picking up the correct exit code (which it might be able to do, even when the exit code is to be ignored due to the target_file) - we should probably verify all these corner cases.

Comment (berland) [reply tests/ert/unit_tests/forward_model_runner/test_forward_model_step.py]: Maybe I would like to see a parameter to the test which tells me if Ert regard the step execution as a success or not.

File: src/_ert/forward_model_runner/forward_model_step.py [hunk 6]
```diff
@@ -226,14 +229,19 @@ def _run(self) -> Generator[Start | Exited | Running]:
...
+            fm_step_pids,
         )
         yield exited_message
 
     def _create_exited_message_based_on_exit_code(
         self,
         max_memory_usage: int,
-        target_file_mtime: int | None,
+        target_file: str | None,
+        existing_target_file_mtime: int | None,
```
Comment (xjules) [inline src/_ert/forward_model_runner/forward_model_step.py:244]: why we need this mtime parameter there?

Comment (jonathan-eq) [reply src/_ert/forward_model_runner/forward_model_step.py:244]: It is being passed to `_check_target_file_is_written`, which then checks if it has been modified by the forward model step.

File: tests/ert/unit_tests/forward_model_runner/test_forward_model_step.py [hunk 7]
```diff
@@ -295,3 +295,128 @@ def test_processtree_timer(
...
+    [
+        pytest.param(0, None, id="exit_code_0_and_target_file_created"),
+        pytest.param(
+            1,
+            "Process exited with status code 1",
+            id="exit_code_1_and_target_file_is_ignored",
+        ),
+    ],
+)
+def test_missing_target_file_sets_error_message_if_specified_and_fm_step_succeeded(
```
Comment (xjules) [inline tests/ert/unit_tests/forward_model_runner/test_forward_model_step.py]: is the name of the function correct - ` _missing_target_file`?

Comment (jonathan-eq) [reply tests/ert/unit_tests/forward_model_runner/test_forward_model_step.py]: I needed it more specific, but ran into ruff formatting issues. This seemed like the best of both world.

Comment (jonathan-eq) [reply tests/ert/unit_tests/forward_model_runner/test_forward_model_step.py]: Do you have suggestions for a better name? It was difficult to find something under the max line length..

Comment (xjules) [reply tests/ert/unit_tests/forward_model_runner/test_forward_model_step.py]: What I meant whether the name is actually correct since in this test the target file is being created, or?

Comment (jonathan-eq) [reply tests/ert/unit_tests/forward_model_runner/test_forward_model_step.py]: You are right, it is the incorrect name 🫤

File: tests/ert/unit_tests/forward_model_runner/test_forward_model_step.py [hunk 8]
```diff
@@ -295,3 +295,128 @@ def test_processtree_timer(
...
+            id="exit_code_0_and_target_file_not_created_gives_error",
+        ),
+        pytest.param(
+            1,
+            "Process exited with status code 1",
+            id="exit_code_1_and_target_file_is_ignored",
+        ),
+    ],
+)
+def test_target_file_sets_error_if_specified_but_not_created_and_fm_step_succeeded(
```
Comment (xjules) [inline tests/ert/unit_tests/forward_model_runner/test_forward_model_step.py]: can I merge (via parametrize) this test and test_target_file_sets_error_if_specified_but_not_updated_and_fm_step_succeeded ?

Comment (jonathan-eq) [reply tests/ert/unit_tests/forward_model_runner/test_forward_model_step.py]: Yes, that should be fine. We can probably merge all of the tests for target_file

**Comment by berland** [comment]
Removed backport label to 14.3, as this is a bug that has become a feature.

Can be backported to 14.4 since this is not released in Komodo yet. Warrants a notice in komodo release notes, as this bugfix might imply that Ert configs that worked last month now crashes (if there has been a target file defined, but not written to, but forward model was fine except that).
